oclif/oclif
# Allow specifying an additional cli for installers

## What does this PR do?
* Adds a hidden flag, `--additional-cli`, to the `oclif pack:win` and `oclif pack:macos` commands that will make an additional Oclif CLI available on the user's machine. In order for this to work correctly the following needs to be true:
    * There needs to be an executable named after the additional CLI in the current CLI's `./bin` directory (a `.cmd` file for Windows and a shell script for Mac).
    * The additional CLI needs to be added at the root of the current CLI's tarball in a folder named after itself.
        * See [`include-sf.js`](https://github.com/salesforcecli/sfdx-cli/blob/main/scripts/include-sf.js) for an example of a [pretarball script](https://github.com/oclif/oclif/blob/master/src/tarballs/build.ts#L70) that does this.
* Fixes several bugs with how tarballs and installers were being uploaded. The tarball config version and package version were different and weren't being used appropriately when dealing with `prerelease` versions.
* Updates CI to use non-outdated versions of node.
* Allows running npm/yarn scripts on Windows with `shx`.

## Testing Instructions
1. Uninstall all versions of the CLI on your machine using the instructions [here](https://developer.salesforce.com/docs/atlas.en-us.234.0.sfdx_setup.meta/sfdx_setup/sfdx_setup_uninstall.htm).
1. Download the testing installer using the AWS CLI:
    1. Mac: 
    ```
    aws s3 cp s3://dfc-data-production/media/salesforce-cli/sfdx/versions/7.126.0-dev.8/47dc692/sfdx-v7.126.0-dev.8-47dc692.pkg ./
    ```
    1. Windows:
    ```
    aws s3 cp s3://dfc-data-production/media/salesforce-cli/sfdx/versions/7.126.0-dev.8/47dc692/sfdx-v7.126.0-dev.8-47dc692-x64.exe ./
    ```
    1. Review [the PR that generated those installers](https://github.com/salesforcecli/sfdx-cli/pull/329).
1. Run the installer.
1. Check that both `sfdx` and `sf` are available (you might need to restart your terminal for env vars to take effect).
    1. Versions should be:
        1. sfdx - `7.126.0-dev.8`
        1. sf - `1.2.0`
        1. Using node `16.13.0`
1. Run `sfdx update dev`
    1. Check CLIs are available as above.
1. Run `sfdx update stable`
    1. Check CLIs are available as above (`sf` will still be available because files aren't deleted with `update` just overwritten).
1. Run `sfdx update dev` again.
    1. Check CLIs are available as above.
